Implement hypot

muthuArivoli opened the pull request:
Related to #38349 
Closes #22764

muthuArivoli committed:
implement hypot

muthuArivoli committed:
add tests and docs

kshitij12345 commented on tools/autograd/derivatives.yaml:
```diff
@@ -514,6 +514,10 @@
   self: hardswish_backward(grad, self)
 
+- name: hypot(Tensor self, Tensor other) -> Tensor
+  self: grad * self / sqrt(self * self + other * other)
```
Great Job! 
Just a passing suggestion. 
You can probably use `result` instead of recomputing with `sqrt(self * self + other * other)`

Example: 
https://github.com/pytorch/pytorch/blob/be19bda57e9f6cdce86bb81309dacf3cea1c48f3/tools/autograd/derivatives.yaml#L424-L428

muthuArivoli committed:
fix cuda name

muthuArivoli committed:
use result in autograd

muthuArivoli commented:
@mruberry PTAL. Not sure why that single build is failing.

Also, I have a few questions:
1. When would you implement the in place version of a function, especially for binary functions? The README seemed to indicate that this was legacy and #38384 didn't implement it for logaddexp, but #40651 did implement it for lcm and gcd.
2. When should you implement AVX/Vec256 support for a function? #38384 did implement it for logaddexp, but #38388 did not for arcsinh, arccosh, and arctanh. 
3. In the Vec256 headers, when should you use the sleef function vs the _mm256 function? I noticed while most used the sleef function, some used the _mm256 function even though there was an equivalent sleef function (like for sqrt, floor, ceil).

Mainly asking these to try to be consistent. Thanks!

mruberry commented:
It looks like the build failed with an infrastructure error. Sorry about that. It doesn't indicate anything is wrong with this PR.

> 1. When would you implement the in place version of a function, especially for binary functions? The README seemed to indicate that this was legacy and #38384 didn't implement it for logaddexp, but #40651 did implement it for lcm and gcd.

We're not very stringent about what is/isn't a method, and it's OK to not implement a method variant of hypot. The [README](https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native) suggests it's "core" operations which are also methods and that functions which are "too complicated" aren't. 

> 2. When should you implement AVX/Vec256 support for a function? #38384 did implement it for logaddexp, but #38388 did not for arcsinh, arccosh, and arctanh.

Adding vectorized CPU support is great and of interest when an operation is part of a performance bottleneck, but not required for the first version of an op (unless it's perf critical).

> 3. In the Vec256 headers, when should you use the sleef function vs the _mm256 function? I noticed while most used the sleef function, some used the _mm256 function even though there was an equivalent sleef function (like for sqrt, floor, ceil). 

That is an excellent question! I think we prefer using the _mm256 where available.

mruberry commented on aten/src/ATen/cpu/vec256/vec256_float_neon.h:
```diff
@@ -349,6 +349,16 @@ template <> class Vec256<float> {
     }
     return loadu(tmp);
   }
+  Vec256<float> hypot(const Vec256<float> &b) const {
```
Since we don't have a great way to test this, maybe let's exclude the neon vectorized version (for now at least)?

mruberry commented on test/test_torch.py:
```diff
@@ -16703,6 +16705,15 @@ def test_remainder_fmod_large_dividend(self, device, dtype):
+    @dtypes(torch.float32, torch.float64)
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    def test_hypot(self, device, dtype):
+        t1 = torch.tensor([3.0, 0, 4.5, -5.5], dtype=dtype, device=device)
```
This is a fine test but you might as well use randn to generate a set of tensors and then compare them with NumPy.

For example:

```
inputs = [
  (torch.randn((3, 3,), device=device, dtype=dtype), torch.randn((3, 3,), device=device, dtype=dtype),
  ...
]
```

You could include a 1D set of inputs and a 3D set of inputs, for example, plus a set of inputs that are transposed.

mruberry replied:
Better also add a test case where the tensor on the lhs in a long tensor, and the tensor on the rhs is a tensor of the specified dtype.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2776,6 +2776,29 @@ def merge_dicts(*dicts):
 
+add_docstr(torch.hypot,
+           r"""
+hypot(input, other, out=None) -> Tensor
```
`input, other, *, out=None`

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2776,6 +2776,29 @@ def merge_dicts(*dicts):
+Args:
+    input (Tensor): the first input tensor
+    other (Tensor): the second input tensor
+    {out}
```
```
   ...
   other (Tensor): the second input tensor

Keyword args:
  {out}
```

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2776,6 +2776,29 @@ def merge_dicts(*dicts):
+
+Given the legs of a right triangle, return its hypotenuse.
+
+Equivalent to calculating pointwise :math:`\sqrt{x^2 + y^2}`.
```
This should probably use a treatment more like here:

https://github.com/pytorch/pytorch/blob/a7c04774f992c20087ff93e4858bf8f5c7ec01f4/torch/_torch_docs.py#L140

`x` and `y` don't refer to anything in this context, so you'll want to reference `input` and `other`.

mruberry commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -738,6 +738,19 @@ void lcm_kernel(TensorIterator& iter) {
 }
 
+void hypot_kernel(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "hypot_cpu", [&]() {
```
The PR adds a vectorized version of bfloat16 hypot but this dispatch would only be for float32 and float64, right?

mruberry commented on aten/src/ATen/native/BinaryOps.cpp:
```diff
@@ -884,6 +885,21 @@ Tensor& lcm_(Tensor& self, const Tensor& other) {
+}
+
+Tensor hypot(const Tensor& self, const Tensor& other) {
+  Tensor result = at::empty({0}, self.options());
```
This has to be constructed as just `Tensor result;` because it's a binary op which, by default, participates in type promotion. If self has dtype torch.float32, for example, and other has dtype torch.float64, then the result type should be torch.float64 when not specified by the user. This would tell TensorIterator that the result type should be torch.float32, however.

mruberry reviewed (changes requested):
This is a very well written and mostly complete PR, thank you for submitting it, @muthuArivoli!

I suggested a few small corrections and updates.

muthuArivoli committed:
fixes for review

muthuArivoli committed:
fix docs

muthuArivoli committed:
use randint instead of ones

muthuArivoli commented:
@mruberry PTAL, I made all the requested changes

mruberry commented on test/test_torch.py:
```diff
@@ -16853,6 +16855,20 @@ def test_remainder_fmod_large_dividend(self, device, dtype):
                 else:
                     self.assertTrue(abs(c[0] - d[0]) == abs(b[0]))  # differ by one divisor
 
+    @dtypes(torch.float32, torch.float64)
```
better add a test for bfloat16 on CPU:

```
@dtypesIfCPU(torch.bfloat16, torch.float32, torch.float64)
@dtypes(torch.float32, torch.float64)
...
```

mruberry reviewed (approved):
Awesome! Nice work @muthuArivoli. Please add the requested bfloat16 test variant on CPU and then I'll merge this. 

Update: you'll also have to merge with master to rebase past the issue causing several builds to fail. Sorry about that.

muthuArivoli committed:
add test for bfloat16

muthuArivoli committed:
fix bfloat16 dtype and numpy test

muthuArivoli commented:
@mruberry, I added bfloat16 to the tests, but since bfloat16 does not seem to be supported for numpy arrays, I tested against the equivalent formula just for bfloat16. Is that okay, or should I create a different test with hardcoded values just for bfloat16?

mruberry commented:
> @mruberry, I added bfloat16 to the tests, but since bfloat16 does not seem to be supported for numpy arrays, I tested against the equivalent formula just for bfloat16. Is that okay, or should I create a different test with hardcoded values just for bfloat16?

Sounds good. Another option is to compare against fp32 in NumPy.

I'll run this PR through our internal tests now. Let me know if you're interested in implementing another NumPy function.

mruberry commented:
This is failing on some internal mobile builds:

```
error: no member named 'hypot' in namespace 'std'; did you mean simply 'hypot'?
      ret[i] = std::hypot(values[i], b[i]);
```

@dreiss Do you know how we handle mobile issues like this?

muthuArivoli commented:
I realized that I didn't add hypot to `aten/src/ATen/core/aten_interned_strings.h`. Could that be the reason why the build is failing?

If not, should I just remove the Vec256 support, and leave that for a later PR?

mruberry commented:
> I realized that I didn't add hypot to `aten/src/ATen/core/aten_interned_strings.h`. Could that be the reason why the build is failing?
> 
> If not, should I just remove the Vec256 support, and leave that for a later PR?

I'm not sure, unfortunately. This is also happening on nextafter, strangely. I'm going to try to follow-up with the mobile team tomorrow but versions of these PRs without the vectorized code paths would be the fastest way to get them in. We can follow-up with the vectorized code paths after the mobile issues are resolved.

mruberry commented:
Big thank you to @dreiss who took the time to discuss the Android build issues with me offline. 

@dreiss suggests trying two things. First, implement hypot (and nextafter) in c10/util/math_compat.h:

https://github.com/pytorch/pytorch/blob/c660d2a9aea49d0fd2d2ce952b5c2aec3278d853/c10/util/math_compat.h#L47

Second, add a simple vectorized implementation on Neon that just loops over its inputs and calls std::hypot (std::nextafter). I realize I suggested removing the vectorized Neon you wrote previously since it wasn't required for server builds, so my mistake there. @dreiss suggested looking at the implementations of pow and atan2 in that file as a guide, if needed.

The c10/util/match_compat.h implementations for hypot (and nextafter) don't have to work, just like the nexttoward implementation.

Does that make sense? If that doesn't fix the issue then I'll follow-up with @dreiss and the mobile team again and see if we can help out instead of requesting all these changes from you, @muthuArivoli. We appreciate your patience and diligence dealing with these issues.

muthuArivoli committed:
Merge remote-tracking branch 'upstream/master' into implement-hypot

muthuArivoli committed:
mobile fixes

muthuArivoli commented:
@mruberry I added the vectorized implementation back to Neon, and I added the function to the math_compat header. Currently, I'm just throwing an error, but hopefully someone will eventually be able to implement and test that. If this works, I'll go ahead and do the same on the nextafter PR.

mruberry commented on c10/util/math_compat.h:
```diff
@@ -51,6 +51,14 @@ namespace std {
   }
 
+  // TODO: this function needs to be implemented and tested. Currently just throw an error.
+  inline float hypot(float x, long double y) {
```
hypot's signature is (float, float) and (double, double)

mruberry commented:
> @mruberry I added the vectorized implementation back to Neon, and I added the function to the math_compat header. Currently, I'm just throwing an error, but hopefully someone will eventually be able to implement and test that. If this works, I'll go ahead and do the same on the nextafter PR.
> 
> Also, seems like the linting error is from someone else's work.

I'll run the update through the Android builds now.

muthuArivoli committed:
change method signature

mruberry commented:
Android build is fixed. OSS PR build failures look like a failure in the base, not a problem with this revision. Starting the land process now.

mruberry commented:
Looks like a conflict slipped in. Would you please rebase (and fix the nit, I guess, since the PR must be rebased anyway).

muthuArivoli committed:
Merge remote-tracking branch 'upstream/master' into implement-hypot

muthuArivoli commented:
Just fixed the merge conflict and fixed the method signature
